File: packages/modules/lodash/core/typescript/src/lang.ts
/* eslint-disable @typescript-eslint/no-explicit-any */

export function clone<T>(value: T): T {
  if (value === null || typeof value !== 'object') return value;
  if (Array.isArray(value)) return [...value] as unknown as T;
  return { ...value } as T;
}

export function cloneDeep<T>(value: T): T {
  if (value === null || typeof value !== 'object') return value;
  if (value instanceof Date) return new Date(value.getTime()) as T;
  if (value instanceof RegExp)
    return new RegExp(value.source, value.flags) as T;
  if (Array.isArray(value))
    return value.map((item) => cloneDeep(item)) as unknown as T;
  const result: Record<string, any> = {};
  for (const key in value) {
    if (Object.prototype.hasOwnProperty.call(value, key)) {
      result[key] = cloneDeep((value as Record<string, any>)[key]);
    }
  }
  return result as T;
}

export function eq(value: any, other: any): boolean {
  return value === other || (value !== value && other !== other);
}

export function gt(value: any, other: any): boolean {
  return value > other;
}

export function gte(value: any, other: any): boolean {
  return value >= other;
}

export function isArguments(value: any): boolean {
  return Object.prototype.toString.call(value) === '[object Arguments]';
}

export function isArray(value: any): value is any[] {
  return Array.isArray(value);
}

export function isBoolean(value: any): value is boolean {
  return typeof value === 'boolean';
}

export function isDate(value: any): value is Date {
  return value instanceof Date;
}

export function isElement(value: any): boolean {
  return value !== null && typeof value === 'object' && value.nodeType === 1;
}

export function isEmpty(value: any): boolean {
  if (value === null || value === undefined) return true;
  if (Array.isArray(value) || typeof value === 'string')
    return value.length === 0;
  if (typeof value === 'object') return Object.keys(value).length === 0;
  return true;
}

export function isEqual(a: any, b: any): boolean {
  if (a === b) return true;
  if (Number.isNaN(a) && Number.isNaN(b)) return true;
  if (a == null || b == null) return a === b;
  if (typeof a !== typeof b) return false;
  if (typeof a !== 'object') return a === b;
  if (a instanceof Date && b instanceof Date)
    return a.getTime() === b.getTime();
  if (a instanceof RegExp && b instanceof RegExp)
    return a.toString() === b.toString();
  if (Array.isArray(a) && Array.isArray(b)) {
    if (a.length !== b.length) return false;
    for (let i = 0; i < a.length; i++) {
      if (!isEqual(a[i], b[i])) return false;
    }
    return true;
  }
  if (Array.isArray(a) !== Array.isArray(b)) return false;
  const keysA = Object.keys(a);
  const keysB = Object.keys(b);
  if (keysA.length !== keysB.length) return false;
  for (const key of keysA) {
    if (!Object.prototype.hasOwnProperty.call(b, key)) return false;
    if (!isEqual(a[key], b[key])) return false;
  }
  return true;
}

export function isError(value: any): value is Error {
  return value instanceof Error;
}

export function isFunction(value: any): value is (...args: any[]) => any {
  return typeof value === 'function';
}

export function isInteger(value: any): value is number {
  return Number.isInteger(value);
}

export function isLength(value: any): boolean {
  return Number.isInteger(value) && value >= 0;
}

export function isMap(value: any): value is Map<any, any> {
  return value instanceof Map;
}

export function isNaN(value: any): boolean {
  return Number.isNaN(value);
}

export function isNil(value: any): value is null | undefined {
  return value === null || value === undefined;
}

export function isNull(value: any): value is null {
  return value === null;
}

export function isNumber(value: any): value is number {
  return typeof value === 'number';
}

export function isObject(value: any): value is object {
  return (
    value !== null && (typeof value === 'object' || typeof value === 'function')
  );
}

export function isObjectLike(value: any): value is object {
  return value !== null && typeof value === 'object';
}

export function isPlainObject(value: any): boolean {
  if (value === null || typeof value !== 'object') return false;
  const proto = Object.getPrototypeOf(value);
  return proto === Object.prototype || proto === null;
}

export function isRegExp(value: any): value is RegExp {
  return value instanceof RegExp;
}

export function isSet(value: any): value is Set<any> {
  return value instanceof Set;
}

export function isString(value: any): value is string {
  return typeof value === 'string';
}

export function isSymbol(value: any): value is symbol {
  return typeof value === 'symbol';
}

export function isUndefined(value: any): value is undefined {
  return value === undefined;
}

export function lt(value: any, other: any): boolean {
  return value < other;
}

export function lte(value: any, other: any): boolean {
  return value <= other;
}

export function toArray<T>(value: T | T[]): T[] {
  if (value === null || value === undefined) return [];
  if (Array.isArray(value)) return value;
  if (typeof value === 'string') return value.split('') as unknown as T[];
  if (typeof value === 'object') return Object.values(value) as T[];
  return [];
}

export function toFinite(value: any): number {
  const num = Number(value);
  if (isNaN(num)) return 0;
  if (num === Infinity || num === -Infinity) {
    return num > 0 ? Number.MAX_VALUE : -Number.MAX_VALUE;
  }
  return num;
}

export function toInteger(value: any): number {
  return Math.floor(toFinite(value));
}

export function toNumber(value: any): number {
  return Number(value);
}

export function toString(value: any): string {
  if (value === null || value === undefined) return '';
  if (typeof value === 'string') return value;
  return String(value);
}

File: packages/modules/lodash/core/swift/src/lang.swift
import Foundation

/**
 * Performs a comparison between two values to determine if they are equivalent.
 */
public func eq(_ value: Any?, _ other: Any?) -> Bool {
    if value == nil && other == nil { return true }
    guard let v = value, let o = other else { return false }
    // Basic Equatable check is hard with Any. Using string representation as a fallback MVP.
    return "\(v)" == "\(o)"
}

/**
 * Checks if value is greater than other.
 */
public func gt(_ value: Double, _ other: Double) -> Bool {
    return value > other
}

/**
 * Checks if value is greater than or equal to other.
 */
public func gte(_ value: Double, _ other: Double) -> Bool {
    return value >= other
}

/**
 * Checks if value is less than other.
 */
public func lt(_ value: Double, _ other: Double) -> Bool {
    return value < other
}

/**
 * Checks if value is less than or equal to other.
 */
public func lte(_ value: Double, _ other: Double) -> Bool {
    return value <= other
}

/**
 * Checks if value is likely an array.
 */
public func isArray(_ value: Any?) -> Bool {
    return value is [Any]
}

/**
 * Checks if value is classified as a boolean primitive or object.
 */
public func isBoolean(_ value: Any?) -> Bool {
    return value is Bool
}

/**
 * Checks if value is empty.
 */
public func isEmpty(_ value: Any?) -> Bool {
    if value == nil { return true }
    if let str = value as? String { return str.isEmpty }
    if let arr = value as? [Any] { return arr.isEmpty }
    if let dict = value as? [String: Any] { return dict.isEmpty }
    return true
}

/**
 * Performs a deep comparison between two values to determine if they are equivalent.
 */
public func isEqual(_ a: Any?, _ b: Any?) -> Bool {
    return eq(a, b)
}

/**
 * Checks if value is a finite primitive number.
 */
public func isFinite(_ value: Any?) -> Bool {
    guard let n = value as? Double else { return false }
    return n.isFinite
}

/**
 * Checks if value is classified as a Number primitive or object.
 */
public func isNumber(_ value: Any?) -> Bool {
    return value is Double || value is Int || value is Float
}

/**
 * Checks if value is classified as a String primitive or object.
 */
public func isString(_ value: Any?) -> Bool {
    return value is String
}

/**
 * Checks if value is NaN.
 */
public func isNaN(_ value: Any?) -> Bool {
    guard let n = value as? Double else { return false }
    return n.isNaN
}

/**
 * Checks if value is null.
 */
public func isNil(_ value: Any?) -> Bool {
    return value == nil
}

/**
 * Checks if value is null.
 */
public func isNull(_ value: Any?) -> Bool {
    return value == nil
}

/**
 * Checks if value is an integer.
 */
public func isInteger(_ value: Any?) -> Bool {
    if let n = value as? Int { return true }
    if let n = value as? Double { return n.truncatingRemainder(dividingBy: 1.0) == 0.0 }
    return false
}

/**
 * Converts value to a finite number.
 */
public func toFinite(_ value: Any?) -> Double {
    let n = toNumber(value)
    if n.isInfinite {
        return n > 0 ? Double.greatestFiniteMagnitude : -Double.greatestFiniteMagnitude
    }
    return n.isNaN ? 0.0 : n
}

/**
 * Converts value to an integer.
 */
public func toInteger(_ value: Any?) -> Int {
    return Int(toFinite(value))
}

/**
 * Converts value to a number.
 */
public func toNumber(_ value: Any?) -> Double {
    if let n = value as? Double { return n }
    if let n = value as? Int { return Double(n) }
    if let s = value as? String { return Double(s) ?? 0.0 }
    if let b = value as? Bool { return b ? 1.0 : 0.0 }
    return 0.0
}

/**
 * Converts value to a string.
 */
public func toString(_ value: Any?) -> String {
    guard let value = value else { return "" }
    return "\(value)"
}

/**
 * Casts value as an array if it's not one.
 */
public func castArray(_ value: Any?) -> [Any] {
    if let arr = value as? [Any] { return arr }
    return [value as Any]
}

/**
 * Checks if value is classified as a Function object.
 */
public func isFunction(_ value: Any?) -> Bool {
    // Hard in Swift to check if 'Any' is a closure type generally.
    return false 
}

/**
 * Checks if value is classified as a Date object.
 */
public func isDate(_ value: Any?) -> Bool {
    return value is Date
}

/**
 * Checks if value is an Error object.
 */
public func isError(_ value: Any?) -> Bool {
    return value is Error
}
